content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+It is shown that while a histogram approach loses resolution (i.e., it may not be able to provide the latency value at an arbitrary percentile point), it is far more efficient, and that the cost of only being able to provide the fractions of the population with values _below a predetermined set of values_ is a sensible choice in the context of addressing the question of meeting a performance SLA target.
+## Terms and Definitions
+
+__**Latency**__  
+The time elapsed between the start_time and end_time of some event.
+
+__**Histogram**__  
+A division of a dimension of values into bins defined over continuous ranges, and the associated size of the population whose values fall in each bin.  
+For example, for a dimension of numbers with values between 1 and 100, a histogram with the bins 1-50 and 51-100, and the population {1,2,3,95}, the histogram would show  
+&nbsp;&nbsp;&nbsp;&nbsp;bin 1-50:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3 members  
+&nbsp;&nbsp;&nbsp;&nbsp;bin 51-100:&nbsp;&nbsp;&nbsp;&nbsp;1 members
+
+__**Percentile**__  
+The value below which a given percentage of observations in a group of observations fall.  
+For example, the 75th percentile of {1,2,3,4} is 3.  
+Likewise, the 75th percentile of {1,4,3,1} is 3.  
+That is, percentiles are applied to _value-ordered populations_.
+
+__**SLA**__  
+Service Level Agreement.  
+e.g. "99% of events will be processed in less than 1ms."
+
+__**Worker**__  
+A Wallaroo worker process. Part of a Wallaroo application cluster.
+
+__**Pipeline**__  
+A Wallaroo application pipeline. Part of a Wallaroo application topology. A pipeline consists of computations (stateless or stateful), partitions, sources, and (optionally) sinks.
+
+__**Step**__  
+A step is a logical unit in a Wallaroo pipeline which is a part of a larger topology. Any individual component of a pipeline is a step for the purposes of monitoring.
+
+__**Boundary**__  
+A part Wallaroo's internal topology representation that is responsible for networking between separate workers. Ingress and egress boundaries make natural points on which to measure worker throughput.
+
+__**Source**__  
+The first step in any Wallaroo pipeline, which is responsible for receiving incoming data from an external system such as a TCP sender, Kafka, etc.
+
+__**Sink**__  
+The last step in a Wallaroo pipeline which emits output, which is responsible for sending outgoing data from Wallaroo to an external system, such as a TCP receiver, Kafka, etc.
+
+__**Metrics Receiver**__  
+A instance on each Wallaroo worker that is responsible for collecting and processing metrics produces on the worker. The same instance is also responsible for emitting the aggregate collection to an external monitoring system.
+
+## Problem Statement
+
+We would like to provide our clients with an SLA stating that 99% of the events we process have an end-to-end latency of less than 1ms, and 99.9% have an end-to-end latency of less than 1s.
```

Comment:
Same comment as earlier about the space for `1ms`.

Also, we say `99.9% have an end-to-end latency of less than 1s` here but earlier we say `99.9% of events are processed in less than 10ms`. These should probably be the same threshold.